oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/annotation/OAuth2Client.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client.annotation;
+
+import org.springframework.core.annotation.AliasFor;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.web.method.annotation.OAuth2ClientArgumentResolver;
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * This annotation may be used to resolve a method parameter into an argument value
+ * for the following types: {@link ClientRegistration}, {@link OAuth2AuthorizedClient}
+ * and {@link OAuth2AccessToken}.
+ *
+ * <p>
+ * For example:
+ * <pre>
+ * &#64;Controller
+ * public class MyController {
+ *     &#64;GetMapping("/client-registration")
+ *     public String clientRegistration(@OAuth2Client("login-client") ClientRegistration clientRegistration) {
+ *         // do something with clientRegistration
+ *     }
+ *
+ *     &#64;GetMapping("/authorized-client")
+ *     public String authorizedClient(@OAuth2Client("login-client") OAuth2AuthorizedClient authorizedClient) {
+ *         // do something with authorizedClient
+ *     }
+ *
+ *     &#64;GetMapping("/access-token")
+ *     public String accessToken(@OAuth2Client("login-client") OAuth2AccessToken accessToken) {
+ *         // do something with accessToken
+ *     }
+ * }
+ * </pre>
+ *
+ * @author Joe Grandja
+ * @since 5.1
+ * @see OAuth2ClientArgumentResolver
+ */
+@Target(ElementType.PARAMETER)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+public @interface OAuth2Client {
```

Comment:
I do like `@OAuth2Client` as well. It just feels and looks natural. However, I can appreciate your concern regarding the possibility that one annotation may not be future proof enough. It can be risky for sure.

This is what we have so far:

```
public String clientRegistration(@OAuth2Client("github") ClientRegistration clientRegistration) {

public String authorizedClient(@OAuth2Client("github") OAuth2AuthorizedClient authorizedClient) {

public String accessToken(@OAuth2Client("github") OAuth2AccessToken accessToken) {
```

Here is an alternative suggestion:

```
public String authorizedClient(@OAuth2AuthorizedClient("github") OAuth2AuthorizedClient authorizedClient) {

public String accessToken(@OAuth2AuthorizedClient("github") OAuth2AccessToken accessToken) {
```

NOTE: `ClientRegistration` is contained within `OAuth2AuthorizedClient`  so I feel we can drop this from being resolved.


For HTTP Client integration, which is coming up next, this is what I'm proposing:

```
	@OAuth2Client("github")
	private RestTemplate githubClient;

	@OAuth2Client("github")
	private WebClient githubClient;
```

I feel `@OAuth2Client` is natural here and should stick over `@OAuth2AuthorizedClient` in this instance.


> Can you think of anything an oauth2 client would want resolved for it in the future...

Possibly a `RefreshToken`,  `IDToken`, `Jwt`, or `IntrospectedAccessToken`. I'll have to give this some more thought as I can't think of anything else at the moment.

Reply:
I've thought about this further and I feel we should avoid the single annotation pattern.

My vote goes for:

`@OAuth2AuthorizedClient` resolves `OAuth2AuthorizedClient` or `OAuth2AccessToken`

`@OAuth2Client` will serve as a _Stereotype_ annotation acting in the role of the OAuth 2.0 Client. This would be used to annotate the HTTP Client integrations, for example, `RestTemplate`, `WebClient`, etc.

What are your thoughts on this approach @rwinch ?

Reply:
Actually `@OAuth2AuthorizedClient` won't work as it collides with `OAuth2AuthorizedClient`. Need to think of a better name. I think  `@AuthorizedClient` is too generic.